Source/DafnyCore/AST/TypeDeclarations/NewtypeDecl.cs

```diff
@@ -96,13 +96,21 @@ public Type RhsWithArgument(List<Type> typeArgs) {
   public TopLevelDecl AsTopLevelDecl => this;
   public TypeDeclSynonymInfo SynonymInfo { get; set; }
 
-  public TypeParameter.EqualitySupportValue EqualitySupport {
+  private IndDatatypeDecl.ES equalitySupport = IndDatatypeDecl.ES.NotYetComputed;
+
+  public IndDatatypeDecl.ES EqualitySupport {
     get {
-      if (this.BaseType.SupportsEquality) {
-        return TypeParameter.EqualitySupportValue.Required;
-      } else {
-        return TypeParameter.EqualitySupportValue.Unspecified;
+      if (equalitySupport == IndDatatypeDecl.ES.NotYetComputed) {
+        var thingsChanged = false;
+        if (ModuleResolver.SurelyNeverSupportEquality(BaseType)) {
```

Comment:
Because the types are topologically sorted, this will only be executed when the types this newtype depends on have been resolved.